Microsoft.Identity.Web/Client/TokenCacheProviders/Redis/MSALAppRedisTokenCacheProviderExtension.cs

```diff
@@ -0,0 +1,86 @@
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+using Microsoft.AspNetCore.Authentication.AzureAD.UI;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using System;
+
+namespace Microsoft.Identity.Web.Client.TokenCacheProviders
+{
+    public static class MSALAppRedisTokenCacheProviderExtension
+    {
+        /// <summary>Adds both the app and per-user redis token caches.</summary>
+        /// <param name="services">The services collection to add to.</param>
+        /// <param name="cacheOptions">the MSALRedisTokenCacheOptions allows the caller to set the token cache expiration configuration options</</param>
+        /// <returns></returns>
+        public static IServiceCollection AddRedisTokenCaches(this IServiceCollection services, MSALRedisTokenCacheOptions cacheOptions = null)
+        {
+            var redisCacheoptions = (cacheOptions == null) ? throw new NullReferenceException("RedisCacheOptions are not provided")
+            : cacheOptions;
+
+            AddRedisAppTokenCache(services, redisCacheoptions);
+            AddRedisPerUserTokenCache(services, redisCacheoptions);
+            return services;
+        }
+
+
+        /// <summary>Adds the distributed token cache to the service collection.</summary>
+        /// <param name="services">The services collection to add to.</param>
+        /// <param name="cacheOptions">the MSALRedisTokenCacheOptions allows the caller to set the token cache expiration and Redis configuration options</param>
+        public static IServiceCollection AddRedisAppTokenCache(this IServiceCollection services, MSALRedisTokenCacheOptions cacheOptions)
+        {
+            services.AddStackExchangeRedisCache(cacheOptions.RedisCacheOptions);
+            services.AddSingleton<IMSALAppTokenCacheProvider>(factory =>
+            {
+                var distributedCache = factory.GetRequiredService<IDistributedCache>();
+                var optionsMonitor = factory.GetRequiredService<IOptionsMonitor<AzureADOptions>>();
+
+                return new MSALAppRedisTokenCacheProvider(distributedCache, cacheOptions, optionsMonitor);
+            });
+
+            return services;
+        }
+
+        /// <summary>Adds the distributed per user token cache to the service collection.</summary>
+        /// <param name="services">The services collection to add to.</param>
+        /// <param name="cacheOptions">the MSALRedisTokenCacheOptions allows the caller to set the token cache expiration and Redis configuration options</param>
+        /// <returns></returns>
+        public static IServiceCollection AddRedisPerUserTokenCache(this IServiceCollection services, MSALRedisTokenCacheOptions cacheOptions)
+        {
```

Comment:
missing validation for a public API input parameter:
`if (cacheOptions == null) throw new ArgumentNullException(nameof(cacheOptions));`